Mai's test changes to coding challenge

Maivue opened the pull request:
In this PR:

Added tests for coding challenge. Usually, I would list out the test plan/cases in the Jira ticket, but for this purpose, I'll document below in the PR. 

**Changes in PR**

- Added user type & login functionality to global setup to allow for multiple user logins
- Updated global teardown to remove auth files after test runs
- Added Login & Inventory spec tests 
- Added Pages (using Page Object Model) to implement Login & Inventory Class for reusability
- Added utility folder to check API responses
- Playwright config updated with storage state
- Added additional values in .env example
- Updated package.json to add simple scripts 
- updated gitIgnore with auth file
- Updated Readme. to include scripts added 


**How am I feeling after???**

Thank you for the opportunity to work on this project. It was super fun and got to learn more Playwright in depth. I would love to hear some feedback on my PR if the team is able to provide feedback. Overall, really fun project to get started in Playwright. 


![Alt Text](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExMWl3amEwYWthNTRkbmJvaDgzaWFjZDB4ZWN4ODNhZjY2c2JyZjJ4cyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/rdma0nDFZMR32/giphy.gif)




**Test Plan**: Login and Inventory Pages Usability
- To validate the usability and functionality of the Login & Inventory pages for a standard and error user. Ensures expected behaviors for positive & negative test scenarios.

**Scope**

1. Pages Covered:

- Login Page
- Inventory Page

2. User Types:

- Standard User 
- Error User
- Invalid User

3. Functional Areas:

- Login functionality
- Inventory management (add/remove items)


**Test Scenarios/Cases**

Login Page

- Standard User & Error User can successfully log in with valid credentials.
- Invalid User is blocked from logging in with invalid credentials and sees the appropriate error message.

Inventory Page

- Standard User can successfully add an item to the cart.
- Standard User can successfully remove an item from the cart.
- Negative case: Running the same Positive test cases with an Error User should fail the inventory tests above.

Maivue committed:
first commit for all test changes

owntheweb commented:
Hello Mai,
Thanks for your super fast PR! If you have time, let's treat this like a general pull request for a project. I have some feedback and a request to continue this coding challenge, with the understanding that Playwright is a new adventure for you and that you asked for feedback. :)

The good:

- Successfully submitted a Git pull request
- Bonus points for including an animated gif in a pull request
- Added tests and settings for multiple user types
- Optimized starting tests with npm run start
- Use of TypeScript seems good and the tests run.
- I like how the tests were organized.

Room for improvement:

- Test count (4) feels a bit light for this coding challenge with it due on the 13th (with understanding that I didn't specify a minimum, I'll take that). As you are focusing on inventory, adding and removing from the cart, perhaps this could be expanded more to test the cart further or anything else you see in that area?
- It will be good to have instructions on how to use the new error and default type users (there was mention of it in the PR and I figured it out by looking through the code). "DEFAULT" === "standard_user" also took a second to work out.
- While testing multiple user types is possible, it currently only tests one type at a time based on .env settings rather than cycling through all the users types in a single test run. In Playwright, it is possible to test under different types of users/roles/browsers in one run. For example, Playwright fixtures with a bit of work can be used to help automate testing different user types (as one option). With this, I'm also open to feedback on how you handled this with previous experiences.
- When testing ERROR type user, a couple of tests are failing as far as I can test. Perhaps they were meant to fail, yet I either missed how to run the test or these tests can be written to successfully fail for this user type without failing the tests themselves?

Unknowns:

- While tests appear to run successfully, I'm seeing this a few times: `Warning: Expected status 200 but received 404 for URL: https://www.saucedemo.com/inventory.html.` Not sure what's causing that at a glance as I can follow the successful steps being completed.

Thanks for taking a look! If there are questions, please let me know.

Maivue committed:
Updated to add for multi-user in parallel

Maivue commented:
Thank you Chris for the feedback. Super appreciate it and great learning process for me. 
So, treating this as a general PR, I did continue the coding challenge a bit further but note this isn't a complete test suite. 😉
I really wanted to update the PR to include multi-user in parallel and here is my solution. 




Addressing the feedback from PR. 

- Test count (4) - Yes, initially there was no specification to minimum test cases, so I figured, this was a good starting point to showcase how I would organize tests, implement in POM and show as a starting point. 
- I did add a few more test cases so you could see the thought process of what tests I would include next, but please note this is not a full test suite. (I wasn't sure how much time to commit to a coding project as many I've done in the past usually ask for no more than 4 hours of project time.  🤷🏻‍♀️)
- I did not remove the global setup but this could be completely removed as its not used in the tests. Figured, I could leave it for reference of the original PR. Furthermore, I did not include instructions since its not used and am open to completely deleting it. 
- The Unknown issue:
My original implementation was looking for a json response and after further investigation the saucedemo page was returning an html instead. I have fixed this and should be returning a status "200" now and warning should no longer appear.  Great catch!


**New Notes on latest commit**
To address the multi users to run in parallel 
- Updated playwright config to include the setup project
- As part of the setup project, I have implemented a auth file, following playwrights documentation for multiple roles. (currently using 3 users) (https://playwright.dev/docs/auth#testing-multiple-roles-together) This allows multiple users to run through the tests at the same time. 
> [!IMPORTANT]  
When using any other users on saucedemo lab to run through the tests, it will fail the inventory tests due to the limitations on their accounts. 
- Updated readme to allow users to understand how to use .env for this project and gave an overview of test cases & test scope
- Updated .env.example to include additional users
- Updated global.teardown to remove all auth files after test runs.
- Added additional cart page to test a some functionality on cart page


![Alt Text](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExYXh6c3V0eWVkaThsdGx6dTd4ZG9vNXd5aHF3Njlmb2FteDR5MWh1MSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/l0MYPkRAqZ3w1PN28/giphy.gif)



Thanks again for the continuation. Let me know if there are any questions.
